File: 103. Binary Tree Zigzag Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if ( !root ) {
        return [];
    }

    var result = [];
    var queue = [];
    queue.push( root );
    var index = 1;

    while ( queue.length > 0 ) {
        var length = queue.length;
        var level = [];

        for ( var i = 0; i < length; i++ ) {
            var node = queue.shift();
            if ( (index + 1) % 2 == 0 ) {
                level.push( node.val );
            } else {
                level.unshift( node.val );
            }

            if ( node.left ) {
                queue.push( node.left );
            }

            if ( node.right ) {
                queue.push( node.right );
            }
        }

        result.push( level );
        index ++;
    }

    return result;
};

File: 103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque


class Solution(object):
    def zigzagLevelOrder(self, root):
        result = []
        if root is None:
            return result
        q = deque()
        q.append(root)
        level = 1

        while q:
            l = len(q)
            tmp = deque()
            for i in xrange(l):
                node = q.popleft()
                if level % 2 == 1:
                    tmp.append(node.val)
                else:
                    tmp.appendleft(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level += 1
            result.append(list(tmp))
        return result
